Heroic Game Launcher repackaging

aidalgol opened the pull request:
###### Description of changes
Repackage the latest stable release of Heroic Games Launcher (HGL) from source instead of the prebuilt AppImage.  This allows us to use dependencies from nixpkgs where available instead of bundled binaries from upstream for everything.

HGL is now also wrapped in a FHS user env using `buildFHSUserEnvBubblewrap`, following the example set by `lutris` and `lutris-unwrapped`.

Add myself as a maintainer for this package (the FHS-wrapped `heroic` package inherits the `meta` attribute` from the `heroic-unwrapped` package).

###### Things done
- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Supersedes #187631 #179608.
Fixes #171077, fixes #177882

I have tested this package with a few native and Windows games.  Native games require enabling the "Use Steam Linux Runtime" option to run at all, but they then run without issue.  Windows games were tested with the third-party [Wine-GE](https://github.com/GloriousEggroll/wine-ge-custom) builds, as that is what I had been using with Lutris (before a recent regression in Lutris) to get Windows games to run reliably on my system.

@WolfangAukang, I apologise for the doubling up of work here, but I had been working on this before you created your PR.  I believe the changes in this PR more fully address the issues in the PRs and tickets mentioned above.

aidalgol committed:
heroic: Repackage from source and wrap in FHS

* Instead of using the prebuilt AppImage, build from source and wrap with
  Electron from nixpkgs.

* Wrap in FHS user environment with buildFHSUserEnvBubblewrap.

* Correct meta.description to mention GOG support.

* Simplify meta attribute

* Update maintainers.
  Taking over sole maintainership from wolfanguakang.

* Use gogdl and legendary-gl from nixpkgs

aidalgol commented:
I'm sorry, I'm going to need some help understanding the ofborg-eval error.

SuperSandro2000 commented on nixos/doc/manual/release-notes/rl-2211.section.md:
```diff
@@ -164,6 +164,8 @@ Use `configure.packages` instead.
 
 - The `pass-secret-service` package now includes systemd units from upstream, so adding it to the NixOS `services.dbus.packages` option will make it start automatically as a systemd user service when an application tries to talk to the libsecret D-Bus API.
 
+- The `heroic` package is now wrapped in a FHS environment so that [third-party Wine builds](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/wiki/How-To:-Wine-and-Proton) (such as Wine-GE) will work.  The bare `heroic` package is named `heroic-unwrapped`.
+
```
Is that so note worthy? I wouldn't have created a changelog entry here.

aidalgol replied:
I was unsure, so I thought it better to add one.  I shall drop it.

SuperSandro2000 commented on pkgs/games/heroic/default.nix:
```diff
@@ -1,39 +1,75 @@
-in
-appimageTools.wrapType2 {
-  inherit name src;
+  packageJSON = "${src}/package.json";
+  yarnLock = "${src}/yarn.lock";
```
Thats IFD and not allowed. You would need to use the FOD variant or vendor the file.

aidalgol replied:
These files are now vendored.

SuperSandro2000 commented on pkgs/games/heroic/default.nix:
```diff
@@ -1,39 +1,75 @@
+
+  DISABLE_ESLINT_PLUGIN = "true";
+
+  postBuild = let
+    yarnFlags = [
+      "--production"
+      "--offline"
+      "--frozen-lockfile"
+      "--ignore-engines"
+      "--ignore-scripts"
+    ];
+  in ''
+    yarn ${lib.escapeShellArgs yarnFlags} --lockfile ${yarnLock} build-electron
+    yarn ${lib.escapeShellArgs yarnFlags} --lockfile ${yarnLock} build
+  '';
```
```suggestion
  postBuild = let
    yarnCmd = "yarn --production --offline --frozen-lockfile --ignore-engines --ignore-scripts --lockfile ${yarnLock}"
  in ''
    ${yarnCmd} build-electron
    ${yarnCmd} build
  '';
```

SuperSandro2000 commented on pkgs/games/heroic/fhsenv.nix:
```diff
@@ -0,0 +1,98 @@
+    zlib
+  ] ++ xorgDeps pkgs;
+
+  meta = {
+    inherit (heroic-unwrapped.meta)
+      homepage
+      description
+      platforms
+      license
+      maintainers
+      broken;
+
+    mainProgram = "heroic";
+  };
```
```suggestion
  meta = heroic-unwrapped.meta // {
    mainProgram = "heroic";
  };
```
also meta should be last.

SuperSandro2000 commented on pkgs/games/heroic/fhsenv.nix:
```diff
@@ -0,0 +1,98 @@
+  targetPkgs = pkgs: with pkgs; [
+    heroic-unwrapped
+
+    gnugrep
+    gnused
+    gnutar
```
Some of these are already included by default https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/build-fhs-userenv-bubblewrap/env.nix#L45-L48

Please remove them to keep the list shorter.

aidalgol replied:
Fixed.

WolfangAukang commented on pkgs/games/heroic/default.nix:
```diff
@@ -1,39 +1,69 @@
     homepage = "https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher";
     license = licenses.gpl3Only;
-    maintainers = with maintainers; [ wolfangaukang ];
+    maintainers = with maintainers; [ wolfangaukang aidalgol ];
```
```suggestion
    maintainers = with maintainers; [ aidalgol ];
```

As the whole package is being restructured, I'd rather have myself removed from the maintainers list.

aidalgol replied:
I'm quite happy to take over looking after this package.  Thank you very much for your initial work in packing this for NixOS. :slightly_smiling_face:

WolfangAukang replied:
Sure thing, in fact I would thank you for the effort of complementing this package with a decent wrapper. I do not play games on my PC anymore, so I can't find any other reason to maintain this package.

WolfangAukang commented on pkgs/games/heroic/default.nix:
```diff
@@ -1,39 +1,69 @@
+}:
+
+mkYarnPackage rec {
+  pname = "heroic-original";
```
```suggestion
  pname = "heroic-unwrapped";
```

aidalgol replied:
I'd like a second opinion on this change.  I went with the name `heroic-original` to be consistent with Lutris, which serves a similar purpose to Heroic, and they are now structured similarly.  I have no preference myself, but I want to be following current conventions in nixpkgs.

WolfangAukang replied:
I threw this review because the name on the `all-packages.nix` is `heroic-unwrapped`, so it is more to keep it with the same name. You can change it to `heroic-original` though, as Lutris does the same, but it is mostly to keep the same name between the `default.nix` and the `all-packages.nix` files.

aidalgol replied:
I did notice that inconsistency with Lutris, so it would not surprise me if a change to the naming of the unwrapped Lutris package comes out of this review.

WolfangAukang replied:
In this case, I would favor naming it as `unwrapped` then, also because Lutris is the only case where it is named as original, but as you said, it uses `unwrapped` on the `all-packages.nix` file.

SuperSandro2000 replied:
Yes, using ``unwrapped`` here would be the standard way I would choose, too.

WolfangAukang commented on pkgs/games/heroic/fhsenv.nix:
```diff
@@ -0,0 +1,104 @@
+
+  targetPkgs = pkgs: with pkgs; [
+    heroic-unwrapped
+
```
```suggestion
```

WolfangAukang commented on pkgs/games/heroic/default.nix:
```diff
@@ -1,39 +1,69 @@
+, fetchFromGitHub
+, makeWrapper
+, electron
+
```
```suggestion
```

WolfangAukang reviewed (changes requested):
Small fixes. Also, can you please squash your commits?

WolfangAukang commented on pkgs/games/heroic/default.nix:
```diff
@@ -1,39 +1,69 @@
+  '';
 
-    install -m 444 -D ${appimageContents}/heroic.desktop -t $out/share/applications
+  doDist = false;
```
It would be good to specify why this is set as false as a future reference.

aidalgol commented:
Changing PR to draft status because I have encountered a major bug.  After sleeping and waking my machine, Heroic launches with a solid grey window and the following error output:
```console
[2:0827/073036.209940:ERROR:gpu_process_host.cc(966)] GPU process exited unexpectedly: exit_code=256
[2:0827/073036.278739:ERROR:gpu_process_host.cc(966)] GPU process exited unexpectedly: exit_code=256
[2:0827/073036.336030:ERROR:gpu_process_host.cc(966)] GPU process exited unexpectedly: exit_code=256
[2:0827/073036.355161:ERROR:gpu_process_host.cc(966)] GPU process exited unexpectedly: exit_code=256
[2:0827/073036.374824:ERROR:gpu_process_host.cc(966)] GPU process exited unexpectedly: exit_code=256
[2:0827/073036.394789:ERROR:gpu_process_host.cc(966)] GPU process exited unexpectedly: exit_code=256
[101:0827/073036.398425:ERROR:gpu_init.cc(486)] Passthrough is not supported, GL is disabled, ANGLE is
```

This appears to be a fairly common issue, and none of the suggested workarounds I have tried work.
- [Open upstream Electron bug report](https://github.com/electron/electron/issues/32317)
- [Electron Reddit thread](https://www.reddit.com/r/electronjs/comments/qdauhu/passthrough_is_not_supported_gl_is_disabled_angle/)

aidalgol commented:
A workaround has been found and implemented in this package.  Ready for review again.

WolfangAukang commented on pkgs/games/heroic/fhsenv.nix:
```diff
@@ -0,0 +1,108 @@
+
+  targetPkgs = pkgs: with pkgs; [
+    heroic-unwrapped
+
```
Can you please remove this extra line? With that, I can throw the approve for this PR.

SuperSandro2000 commented on pkgs/games/heroic/fhsenv.nix:
```diff
@@ -0,0 +1,108 @@
+  '';
+
+  meta = heroic-unwrapped.meta // {
+    mainProgram = "heroic";
```
I think we should move this to heroic because it should already be not correct there.

aidalgol committed:
gogdl: init at 0.3

WolfangAukang commented on pkgs/top-level/all-packages.nix:
```diff
@@ -32876,6 +32880,8 @@ with pkgs;
 
   legendary-gl = python38Packages.callPackage ../games/legendary-gl { };
 
+  gogdl = python38Packages.callPackage ../games/heroic/gogdl.nix { };
```
- `gogdl`, although created by the same people from Heroic, should be separate from the `heroic` package as it is possible to use it standalone. I think it would be better to keep it on its own directory instead.
-  Let's try to keep it on the alphabetical order

WolfangAukang replied:
```suggestion
  gogdl = python3Packages.callPackage ../games/heroic/gogdl.nix { };
```

WolfangAukang reviewed (approved):
LGTM

WolfangAukang commented:
The only thing to consider is squashing your `heroic` commits (be it related to versioning or functionality) into a single commit. The rest (`gogdl` and `legendary-gl`) are fine.

moben commented:
Tested on top of NixOS unstable x86_64, installing and launching games from both stores works.

baduhai commented:
Tested on unstable, same experience as @moben, installing different wine versions also works well.

zhaofengli commented:
`gogdl` now fails to build for me with `ModuleNotFoundError: No module named 'setuptools'`. Not sure which module in the dependency chain removed `setuptools`, but adding it into `propagatedBuildInputs` seems to work.

aidalgol commented:
Rebased onto latest `master` to address merge conflict with `legendary`.

aidalgol commented:
Pushed fix for `gogdl`.  (Thanks @zhaofengli!)

imsofi commented:
When running `nixpkgs-review pr 187960` i hit an issue of `undefined variable 'kdialog'` with `fhsenv.nix` under nixpkgs pointing to `fd54651f5ffb4a36e8463e0c327a78442b26cbe7`.

imsofi commented:
Seems dependencies `curl` `hostname` and `gawk` are missing from the FHS?

<details>
<summary>Log output</summary>

```log
(14:41:09) INFO:    [Legendary]:       Running command: /nix/store/1dczd5hlvaz9vra4cw2150v2v3qnzsm4-heroic-2.4.3/libexec/heroic/deps/heroic/public/bin/linux/legendary --version
/bin/sh: line 1: hostname: command not found
(14:41:09) INFO:    [Legendary]:       Legendary location: /nix/store/1dczd5hlvaz9vra4cw2150v2v3qnzsm4-heroic-2.4.3/libexec/heroic/deps/heroic/public/bin/linux/legendary
(14:41:09) INFO:    [Gog]:             GOGDL location: /nix/store/1dczd5hlvaz9vra4cw2150v2v3qnzsm4-heroic-2.4.3/libexec/heroic/deps/heroic/public/bin/linux/gogdl
(14:41:09) INFO:    [Backend]:         

Heroic Version: 2.4.3 Chopper
Legendary Version:  0.20.27 Dark Energy (hotfix)
OS: Fedora Linux KERNEL: 5.19.12-200.fc36.x86_64 ARCH: x64
CPU: Intel Core™ i3-6100U @2.3 GOVERNOR: powersave
RAM: Total: 15.5 GiB Available: 8.45 GiB
GRAPHICS: 
PROTOCOL: wayland

(14:41:09) INFO:    [Backend]:         User Not Found, removing it from Store
APPIMAGE env is not defined, current application is not an AppImage
LaunchProcess: failed to execvp:
xdg-settings
(14:41:09) WARNING: [Backend]:         Failed to register protocol with OS.
[2:1003/144109.830159:ERROR:browser_main_loop.cc(270)] Gtk: gtk_widget_add_accelerator: assertion 'GTK_IS_ACCEL_GROUP (accel_group)' failed
[2:1003/144109.830376:ERROR:browser_main_loop.cc(270)] Gtk: gtk_widget_add_accelerator: assertion 'GTK_IS_ACCEL_GROUP (accel_group)' failed
[2:1003/144109.830599:ERROR:browser_main_loop.cc(270)] Gtk: gtk_widget_add_accelerator: assertion 'GTK_IS_ACCEL_GROUP (accel_group)' failed
(14:41:09) INFO:    [Backend]:         AreWeAntiCheatYet data downloaded
(14:41:10) INFO:    [Backend]:         Checking for new Heroic Updates
(14:41:12) WARNING: [Backend]:         Error Downloading Winetricks
(14:41:12) INFO:    [DXVKInstaller]:   Updating vkd3d to: vkd3d-proton-2.6
(14:41:12) WARNING: [DXVKInstaller]:   Error when downloading vkd3d Error: Command failed: curl -L https://github.com/bottlesdevs/vkd3d-proton/releases/download/vkd3d-proton-2.6/vkd3d-proton-2.6.tar.gz -o /home/sofi/.config/heroic/tools/vkd3d/vkd3d-proton-2.6.tar.gz --create-dirs
/bin/sh: line 1: curl: command not found

(14:41:12) INFO:    [DXVKInstaller]:   Updating dxvk to: dxvk-1.10.3
(14:41:12) WARNING: [DXVKInstaller]:   Error when downloading dxvk Error: Command failed: curl -L https://github.com/doitsujin/dxvk/releases/download/v1.10.3/dxvk-1.10.3.tar.gz -o /home/sofi/.config/heroic/tools/dxvk/dxvk-1.10.3.tar.gz --create-dirs
/bin/sh: line 1: curl: command not found
```
</details>

aidalgol commented:
I added the missing `gogdl` dependencies to the FHS, but I have no idea why `nixpkgs-review` fails to evaluate the derivation, because I can verify that `kdialog` is a derivation at the top level with,
```console
❯ nix repl .
Welcome to Nix 2.8.1. Type :? for help.

Loading '.'...
Added 17064 variables.

nix-repl> kdia
kdialog   kdiamond
nix-repl> :p kdialog
«derivation /nix/store/sr9i07jqdhs3k5pmykmh04q91mb38bbv-kdialog-22.08.1.drv»
```
And the `heroic` derivation builds successfully.

imsofi commented:
Could you try to run nixpkgs-review? As if i pull in the files manually to my own repo, it also builds successfully. It may be an issue with `nixpkgs-review` itself? Not sure.

aidalgol commented:
I did, and got the same error.  I think it may be a bug in `nixpkgs-review`, given even ofborg is happy with these commits.

soupglasses reviewed (approved):
Tested on hardware. Downloaded Wine from Heroic. Played Satisfactory with said Wine. Everything seems to work as it should. :+1:

imsofi commented:
I am having an issue of the desktop icon being low quality. But i have a feeling this might simply be something on my end (i had heroic installed trough flatpak for testing a bit ago, *this was a bad idea*).

zhaofengli reviewed (approved):
Oops, forgot to approve the PR. It works great and we should get this merged soon. The current `heroic` package in nixpkgs is unusable.

baduhai commented:
A bit off-topic, but how many approvals are required to get something like this merged?

vcunat commented:
Let me cross-reference directly &ndash; the `kdialog` was a genuine problem; see PR https://github.com/NixOS/nixpkgs/pull/195628
